File: windows/tests/RealtimeTranslator.Core.Tests/AudioLossTrackerFixtureTests.cs
using System;
using System.Linq;
using System.Text.Json.Nodes;
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.Realtime;
using Xunit;

namespace RealtimeTranslator.Core.Tests;

public sealed class AudioLossTrackerFixtureTests
{
    public static TheoryData<string> CaseNames
    {
        get
        {
            var data = new TheoryData<string>();
            foreach (var node in SharedFixtures.Load("audio")["loss"]!["cases"]!.AsArray())
            {
                data.Add(SharedFixtures.Text(node!["name"]));
            }

            return data;
        }
    }

    // Given: shared fixture の音声欠落ポリシー
    // When: Core の既定値と比較する
    // Then: フレーム長・再接続閾値・窓・キュー容量が一致する
    [Fact]
    public void PolicyAndConstantsMatchFixture()
    {
        var loss = SharedFixtures.Load("audio")["loss"]!.AsObject();
        var reconnect = loss["reconnect"]!.AsObject();
        var policy = AudioLossPolicy.Default;

        Assert.Equal(SharedFixtures.Number(loss["frameDurationMs"]), policy.FrameDurationMilliseconds);
        Assert.Equal(
            SharedFixtures.Number(reconnect["lostMsThreshold"]),
            policy.ReconnectLostMillisecondsThreshold);
        Assert.Equal(
            SharedFixtures.Number(reconnect["windowMs"]),
            policy.ReconnectWindowMilliseconds);
        Assert.Equal(
            SharedFixtures.Number(loss["sendQueueFrameCapacity"]),
            AudioLossPolicy.SendQueueFrameCapacity);
        Assert.Equal(
            SharedFixtures.Number(loss["taintedSegmentWindowMs"]),
            RealtimeSubtitleAssembler.AudioLossTaintWindow.TotalMilliseconds);
    }

    // Given: fixture のフレーム列と独立した期待値
    // When: AudioLossTracker へ順に再生する
    // Then: 累積 metrics と再接続観測が fixture と一致する
    [Theory]
    [MemberData(nameof(CaseNames))]
    public void ReplayMatchesFixture(string name)
    {
        var item = SharedFixtures.Load("audio")["loss"]!["cases"]!
            .AsArray()
            .OfType<JsonObject>()
            .Single(item => SharedFixtures.Text(item["name"]) == name);
        var tracker = new AudioLossTracker();
        var reconnectAt = SharedFixtures.OptionalNumber(item["expected"]!["reconnectAt"]);
        var observedReconnectAt = (int?)null;

        foreach (var node in item["frames"]!.AsArray())
        {
            var frame = node!.AsObject();
            var atMilliseconds = SharedFixtures.Number(frame["atMs"]);
            var observation = tracker.Observe(
                SharedFixtures.Number(frame["generation"]),
                frame["sequence"]!.GetValue<long>(),
                SharedFixtures.Number(frame["discardedMs"]),
                SharedFixtures.Number(frame["queueWaitMs"]),
                atMilliseconds);

            if (observation.ShouldReconnect)
            {
                observedReconnectAt = atMilliseconds;
            }
        }

        var expected = item["expected"]!.AsObject();
        Assert.Equal(SharedFixtures.Number(expected["droppedFrames"]), tracker.Metrics.DroppedFrames);
        Assert.Equal(SharedFixtures.Number(expected["lostMs"]), tracker.Metrics.LostMilliseconds);
        Assert.Equal(SharedFixtures.Number(expected["lossEvents"]), tracker.Metrics.LossEvents);
        Assert.Equal(
            SharedFixtures.Number(expected["maxQueueWaitMs"]),
            tracker.Metrics.MaxQueueWaitMilliseconds);
        Assert.Equal(reconnectAt, observedReconnectAt);
    }
}

File: RealtimeTranslator/Tests/AudioLossTrackerFixtureTests.swift
import XCTest
@testable import RealtimeTranslator

final class AudioLossTrackerFixtureTests: XCTestCase {
    func testLossCasesMatchSharedContract() throws {
        // Given: 独立した期待値を定義する共有音声欠落fixture
        let fixture = try SharedFixtures.load("audio")
        let loss = try XCTUnwrap(fixture["loss"] as? [String: Any])
        let reconnect = try XCTUnwrap(loss["reconnect"] as? [String: Any])
        let cases = try XCTUnwrap(loss["cases"] as? [[String: Any]])

        // When: 各観測列をtrackerへ再生する
        var policy = AudioLossPolicy.default
        let frameDuration = SharedFixtures.number(loss["frameDurationMs"])
        let threshold = SharedFixtures.number(reconnect["lostMsThreshold"])
        let window = SharedFixtures.number(reconnect["windowMs"])
        policy.frameDurationMilliseconds = frameDuration
        policy.reconnectLostMillisecondsThreshold = threshold
        policy.reconnectWindowMilliseconds = window
        XCTAssertEqual(policy, .default)
        XCTAssertEqual(RealtimeAudioFrameQueue.capacity, SharedFixtures.number(loss["sendQueueFrameCapacity"]))
        XCTAssertEqual(
            RealtimeSubtitleAssembler.audioLossTaintWindow * 1_000,
            Double(SharedFixtures.number(loss["taintedSegmentWindowMs"]))
        )

        for item in cases {
            let frames = try XCTUnwrap(item["frames"] as? [[String: Any]])
            let expected = try XCTUnwrap(item["expected"] as? [String: Any])
            let name = item["name"] as? String ?? "loss case"
            var tracker = AudioLossTracker(policy: policy)
            var reconnectAt: Int?
            for frame in frames {
                let observation = tracker.observe(
                    generation: SharedFixtures.number(frame["generation"]),
                    sequence: SharedFixtures.number(frame["sequence"]),
                    discardedMilliseconds: SharedFixtures.number(frame["discardedMs"]),
                    queueWaitMilliseconds: SharedFixtures.number(frame["queueWaitMs"]),
                    atMilliseconds: SharedFixtures.number(frame["atMs"])
                )
                if observation.shouldReconnect {
                    reconnectAt = SharedFixtures.number(frame["atMs"])
                }
            }

            // Then: metricsと再接続観測がfixture契約と一致する
            XCTAssertEqual(tracker.metrics.droppedFrames, SharedFixtures.number(expected["droppedFrames"]), name)
            XCTAssertEqual(tracker.metrics.lostMilliseconds, SharedFixtures.number(expected["lostMs"]), name)
            XCTAssertEqual(tracker.metrics.lossEvents, SharedFixtures.number(expected["lossEvents"]), name)
            XCTAssertEqual(tracker.metrics.maxQueueWaitMilliseconds, SharedFixtures.number(expected["maxQueueWaitMs"]), name)
            XCTAssertEqual(reconnectAt, SharedFixtures.optionalNumber(expected["reconnectAt"]), name)
        }
    }
}
